### ISSUE TITLE
eof is an insufficient check for remaining entries

### ISSUE THREAD

**Issue by dcjones** [BioJulia/Bio.jl#343]
I ran into an issue working on GFF3 that I think suggests we should make a slight change to our low level IO interface.

Currently it works like this:
```julia
while !eof(reader)
    read!(reader, entry)
    # do something with entry
end
```

But with GFF3  is unusual in that GFF entries can end before the end of the file in the case of trailing FASTA sequences or directives causing this to throw `EOFError` even on valid files. This can fail in other cases, e.g. empty files, trailing comments, etc.

This can be worked around like so, but it's very clumsy
```julia
while true
    try
        read!(reader, entry)
        # do something with entry
    catch err
        if isa(err, EOFError)
            break
        end
        rethrow()
    end
end
```

I suggest we change `read!` to return a Bool indicating whether a entry was successfully read. That would change low-level reading to work like C 
```julia
while read!(reader, entry)
    # do something with entry
end
```

**Comment by Ward9250** [comment]
I think that's a sensible suggestion.

**Comment by bicycle1885** [comment]
I noticed `eof` is not perfect solution for some file formats. But making `read!` to return a boolean value is not my favorite style because other `read!` methods defined in the standard library returns the first argument, which is mutated in the method. So, I defined [the `tryread!` function](https://github.com/BioJulia/Bio.jl/blob/ad78989811179dc1fd314479f45311b90954d986/src/Ragel.jl#L11), which is not a method defined in the standard library and returns a `Nullable` object wrapping a record. If there is no more record available from a stream `tryread!` returns a `Nullable` object without record. This method mimics the behavior of `Base.tryparse`.
